Spigot-Server-Patches/0167-Anti-Xray.patch

```diff
@@ -0,0 +1,997 @@
++        log("Anti X-Ray: " + antiXray);
++        engineMode = getInt("anti-xray.engine-mode", 3);
++        log("\tEngine Mode: " + engineMode + " (" + (engineMode == 1 ? "hidden ores" : engineMode == 2 ? "fake ores" : engineMode == 3 ? "fake ores every third block" : "unknown") + ")");
++        hiddenBlocks = getList("anti-xray.hide-blocks", Arrays.asList(new Object[] {
++            "gold_ore", "iron_ore", "coal_ore", "lapis_ore", "mossy_cobblestone", "obsidian", "chest", "diamond_ore", "redstone_ore", "lit_redstone_ore", "clay", "emerald_ore", "ender_chest"
++        }));
++        log("\tHidden Blocks: " + hiddenBlocks);
++        replaceBlocks = getList("anti-xray.replace-blocks", Arrays.asList(new Object[] {
++            "stone", "planks"
++        }));
++        log("\tReplace Blocks: " + replaceBlocks + " (for engine mode 2 and 3)");
++        maxChunkY = getInt("anti-xray.max-chunk-y", 3);
++        log("\tMax Chunk Y: " + maxChunkY + " (obfuscate up to " + ((maxChunkY + 1) * 16) + " blocks)");
++        asynchronous = getBoolean("anti-xray.asynchronous", true);
++        log("\tAsynchronous: " + asynchronous);
++        neighborsMode = getInt("anti-xray.neighbors-mode", 2);
++        log("\tNeighbors Mode: " + neighborsMode + " (" + (neighborsMode == 1 ? "MC default, sometimes the edges of chunks can't be obfuscated" : neighborsMode == 2 ? "wait until neighbor chunks are loaded" : neighborsMode == 3 ? "load neighbor chunks" : "unknown") + ")");
++        antiXrayInstance = new AntiXray(this);
++    }
+ }
+diff --git a/src/main/java/com/destroystokyo/paper/antixray/AntiXray.java b/src/main/java/com/destroystokyo/paper/antixray/AntiXray.java
+new file mode 100644
+index 0000000..def27cf
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/antixray/AntiXray.java
+@@ -0,0 +1,286 @@
++package com.destroystokyo.paper.antixray;
++
++import java.util.HashSet;
++import java.util.Set;
++import java.util.concurrent.Executors;
++import java.util.concurrent.ExecutorService;
++
++import com.destroystokyo.paper.PaperWorldConfig;
++import net.minecraft.server.Block;
++import net.minecraft.server.BlockPosition;
++import net.minecraft.server.Blocks;
++import net.minecraft.server.Chunk;
++import net.minecraft.server.ChunkSection;
++import net.minecraft.server.DataBits;
++import net.minecraft.server.DataPaletteBlock;
++import net.minecraft.server.IBlockData;
++import net.minecraft.server.PacketDataSerializer;
++import net.minecraft.server.PacketPlayOutMapChunk;
++import net.minecraft.server.World;
++
++public class AntiXray {
++    private static final ExecutorService executorService = Executors.newSingleThreadExecutor();
++    private final boolean antiXray;
++    private final int engineMode;
++    private final int maxChunkY;
++    private final boolean asynchronous;
++    private final int neighborsMode;
++    // Used to keep track of which blocks to obfuscate
++    private final boolean[] obfuscateBlocks = new boolean[Short.MAX_VALUE];
```

Comment:
Quick look over the code tells me this is iterated over. Switch to using bitfields in a long array to better-utilize the CPU cache.